### PR TITLE
Первая версия модели на PyTorch Lightning

### PR DESCRIPTION
**Pull request by saridormi**
### Текущее состояние

Код для обучения модели ([`training/trainer_main.py`](https://github.com/saridormi/commit_message_generation/blob/lightning-dev/training/trainer_main.py)) с флагом [`fast_dev_run`](https://pytorch-lightning.readthedocs.io/en/latest/debugging.html#fast-dev-run) (по 1 батчу на train, val и test) работает без ошибок.

### Основные изменения
В датасете:
* Точно отбрасываются примеры, где диффы и сообщения больше заданной длины (по количеству токенов, которые получаются просто разделением по пробелам)
* Для работы с датасетом используется `LightningDataModule`

В декодере:
* Используется свой [Embedding](https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html) слой
* Вместо прошлого механизма внимания используется [MultiheadAttention из PyTorch](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html)

В модели в целом:
* Для работы с моделью используется `LightningModule`
* На этапе тестирования сейчас используется только greedy подход

В метриках:
* Для подсчета accuracy считается количество равных элементов между тензорами с токенами (padding не учитывается)
* Для подсчета BLEU используется [функция из PyTorch Lightning](https://pytorch-lightning.readthedocs.io/en/latest/metrics.html#bleu-score-func)

### TODO и возможные проблемы

* В декодере после каждого шага модели подается только один предыдущий токен - нужно ли подавать все предыдущие? 
* Стоит ли считать accuracy на тензорах с токенами - длина предсказаний может не совпадать с длиной таргета, также токен может быть частью слова, а не целым?
* Какое значение параметра `reduction` использовать в [loss-е](https://pytorch.org/docs/stable/nn.functional.html?highlight=nll_loss#torch.nn.functional.nll_loss) - 'sum' или 'mean'?
* Реализовать beam search

### CODE REVIEW

**Comment by SokolovYaroslav** [comment]
Поставь меня ревьюером :)

**Comment by SokolovYaroslav** [comment]
для этого мб придется сделать меня контрибьютором репозитория

**Comment by SokolovYaroslav** [comment]
> Код для обучения модели (training/trainer_main.py) с флагом fast_dev_run (по 1 батчу на train, val и test) работает без ошибок.
А в обычном режиме с ошибками или это просто сказано, типа работает?

> Какое значение параметра reduction использовать в loss-е - 'sum' или 'mean'?
О, это интересный вопрос. Я бы предложил сумму. Тогда возможно придётся поподбирать оптимальный learning_rate

**Comment by SokolovYaroslav** [comment]
Стоит завести `requirements.txt` и туда записать необходимые библиотеки с текущей версией (то есть через `==`, а не `>=`). Потому что PL часто ломает зависимости при обновлениях.

File: Config.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            config._CONFIG = pickle.load(config_file)
+            config._CONFIG['DEVICE'] = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    if config['SEED'] is not None:
+        make_reproducible(config['SEED'], config['MAKE_CUDA_REPRODUCIBLE'])
+    if is_test:
+        config.change_config_for_test()
+    return config
+
+
+def make_reproducible(seed: int, make_cuda_reproducible: bool) -> None:
```
Comment (SokolovYaroslav) [inline Config.py:81]: В PL есть `seed_everything`

File: Config.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    random.seed(seed)
+    os.environ['PYTHONHASHSEED'] = str(seed)
+    np.random.seed(seed)
+    torch.manual_seed(seed)
+    if make_cuda_reproducible:
+        torch.backends.cudnn.deterministic = True
+        torch.backends.cudnn.benchmark = False
```
Comment (SokolovYaroslav) [inline Config.py:88]: А вообще я бы предложил переехать на [hydra](http://hydra.cc/docs/intro/#!) для конфигов. Можешь поизучать, если будут вопросы -- пиши :)

Comment (SokolovYaroslav) [reply Config.py:88]: Но там нам не нужно прям кучу функциональности. Просто конфигурации на `yaml` и через декоратор прикрутить его

File: metrics/MyAccuracyMetric.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
+from pytorch_lightning.metrics import Metric
+import torch
+
+
+class MyAccuracyMetric(Metric):
```
Comment (SokolovYaroslav) [inline metrics/MyAccuracyMetric.py:5]: Может без `My` сделать?)))

File: training/trainer_main.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+trainer = pl.Trainer.from_argparse_args(args)
+
+# -----------------------
+#         train         -
+# -----------------------
+trainer.fit(encoder_decoder, dm)
+# -----------------------
+#          test         -
+# -----------------------
+trainer.test(datamodule=dm)
```
Comment (SokolovYaroslav) [inline training/trainer_main.py:35]: Еще из модных штук можно `WandBLogger` взять, который уже есть в PL. Тогда будет удобно делиться результатами запусков и они никуда не пропадут

File: metrics/MyBleuMetric.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+        self.n_gram = n_gram
+        self.add_state("bleu", default=torch.tensor(0.0), dist_reduce_fx="sum")
+        self.add_state("total", default=torch.tensor(0), dist_reduce_fx="sum")
+
+    def update(self, preds, targets):
+        self.bleu += bleu_score(translate_corpus=preds, reference_corpus=targets, n_gram=self.n_gram)
+        self.total += 1
+
+    def compute(self):
+        return self.bleu / self.total
```
Comment (SokolovYaroslav) [inline metrics/MyBleuMetric.py:19]: Может в один файл с Accuracy запихать? `metrics` тип

File: Config.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
+import os
```
Comment (SokolovYaroslav) [inline Config.py:1]: Вообще в питоне же обычно файлы в Снейк кейсе пишут, разве нет?

**Review by SokolovYaroslav — CHANGES_REQUESTED** [review]

**Comment by SokolovYaroslav** [comment]
> Реализовать beam search

Пока предлагаю забить

**Comment by SokolovYaroslav** [comment]
Было бы Мега удобно мне (и всем другим людям, которые захотят воспроизвести обучение) сделать конфигурацию в пайчарме и сохранить её в проекте (если ты сама используешь пайчарм :))

Это можно сделать либо настроив хорошенько .gitignore (по этому поводу можешь писать мне, я спрошу у тех, кто шарит) и добавлять .idea в репо.

Либо я в конфигурации запуска увидел галочку `Store as project file` сверху справа

**Comment by SokolovYaroslav** [comment]
> Стоит ли считать accuracy на тензорах с токенами - длина предсказаний может не совпадать с длиной таргета, также токен может быть частью слова, а не целым?

Ну если токенизировать таргет, то можно и стоит так делать, да. А как иначе? Есть какой-то честный токенизатор, который лучше робертовского?

File: training/trainer_main.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+from model.EncoderDecoderModule import EncoderDecoderModule
+from dataset_utils.CMGDataModule import CMGDataModule
+
+from Config import Config
+
+# -----------------------
+#          init         -
+# -----------------------
+config = Config()
```
Comment (SokolovYaroslav) [inline training/trainer_main.py:13]: Обернуть бы в `if __name__ == "__main__"`

**Comment by SokolovYaroslav** [comment]
> В декодере после каждого шага модели подается только один предыдущий токен - нужно ли подавать все предыдущие?

Только если через аттеншон. А так по одному, да

**Comment by SokolovYaroslav** [comment]
> Вместо прошлого механизма внимания используется MultiheadAttention из PyTorch

Хорошее решение :) Q,K,V используешь как в статье, так что всё правильно

**Comment by saridormi** [comment]
> > Код для обучения модели (training/trainer_main.py) с флагом fast_dev_run (по 1 батчу на train, val и test) работает без ошибок.
> 
>  А в обычном режиме с ошибками или это просто сказано, типа работает?

Да, это просто чтобы быстро понять, что все этапы отработали без проблем

> > Стоит ли считать accuracy на тензорах с токенами - длина предсказаний может не совпадать с длиной таргета, также токен может быть частью слова, а не целым?
> 
> Ну если токенизировать таргет, то можно и стоит так делать, да. А как иначе? Есть какой-то честный токенизатор, который лучше робертовского?

Просто в прошлой версии кода accuracy считалась по уже декодированным строкам, я не очень поняла, есть ли в этом необходимость, подумала, что на токенах может получиться немного другой результат из-за Byte-Pair-Encoding (например, какие-то части слов совпадут, а целые слова нет), но если это не так важно, то получается определенно быстрее :)

> Стоит завести `requirements.txt` и туда записать необходимые библиотеки с текущей версией (то есть через `==`, а не `>=`). Потому что PL часто ломает зависимости при обновлениях.

> Было бы Мега удобно мне (и всем другим людям, которые захотят воспроизвести обучение) сделать конфигурацию в пайчарме и сохранить её в проекте (если ты сама используешь пайчарм :))
> 
> Это можно сделать либо настроив хорошенько .gitignore (по этому поводу можешь писать мне, я спрошу у тех, кто шарит) и добавлять .idea в репо.
> 
> Либо я в конфигурации запуска увидел галочку `Store as project file` сверху справа

Хорошо, добавлю `requirements.txt` и посмотрю, что получается с галочкой `Store as project file` :) 

> А вообще я бы предложил переехать на [hydra](http://hydra.cc/docs/intro/#!) для конфигов. Можешь поизучать, если будут вопросы -- пиши :)

> Еще из модных штук можно `WandBLogger` взять, который уже есть в PL. Тогда будет удобно делиться результатами запусков и они никуда не пропадут

Звучит интересно, попробую разобраться! 

Остальные штуки тогда поправлю.

Спасибо большое за советы!

**Review by SokolovYaroslav — CHANGES_REQUESTED** [review]
По поводу остального кода, в котором содержится основная логика модели: кажется надо как-то вместе засесть и я бы послушал, как всё работает, потому что так очень сложно разобраться сразу во всём

**Comment by SokolovYaroslav** [comment]
Где в итоге `seed_everything` находится?

**Comment by saridormi** [comment]
> По поводу остального кода, в котором содержится основная логика модели: кажется надо как-то вместе засесть и я бы послушал, как всё работает, потому что так очень сложно разобраться сразу во всём

Да, конечно, можем договориться на какое-то удобное время :)

> А почему такая старая версия?

Это вроде последняя версия hydra-core, или я чего-то не понимаю?

> Где в итоге seed_everything находится?

Ой, я забыла про это. Спасибо :)

> Мб сделать все кончики в одном файле? А то так как-то разбросанно и надо бегать по файлам

Думаю, так и правда будет лучше, поменяю


А остальные баги поправлю и добавлю type hints

Спасибо еще раз!

**Comment by SokolovYaroslav** [comment]
> Это вроде последняя версия hydra-core, или я чего-то не понимаю?

Согласен, что-то я не туда посмотрел

**Comment by SokolovYaroslav** [comment]
Может вмерджим?
Кажется надо зафиксировать это, так как мы видимо принципиально ничего не починим :)
